### PR TITLE
Develop/features/DLSV2-104 repurpose layout and error pages

### PR DESCRIPTION
**Pull request by kevwhitt-hee**
This is the work to make the layout and error pages generic across apps.

### CODE REVIEW

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs [hunk 1]
```diff
@@ -0,0 +1,172 @@
+﻿namespace DigitalLearningSolutions.Data.Migrations
+{
+    using FluentMigrator;
+    using FluentMigrator.SqlServer;
+    [Migration(202011101344)]
+    public class CompetencyFrameworkDBChanges : Migration
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:6]: Did you mean to include these changes? Seems unrelated to layout and error pages? 🙂

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                .AddColumn("IsFrameworkDeveloper").AsBoolean().NotNullable().WithDefaultValue(false);
+            //DLSV2-41 implied publish status table:
+            Create.Table("PublishStatus")
+                .WithColumn("ID").AsInt32().NotNullable().PrimaryKey().Identity()
+                .WithColumn("Status").AsString(255).NotNullable();
+            Insert.IntoTable("PublishStatus").Row(new { Status = "Draft" });
+            Insert.IntoTable("PublishStatus").Row(new { Status = "In Review" });
+            Insert.IntoTable("PublishStatus").Row(new { Status = "Published" });
+            //DLSV2-41 new frameworks table and DLSV2-35 with FrameworkConfig field:
+            Create.Table("Frameworks")
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:21]: Unrelated to layout and error pages, but should frameworks really be a separate table to SelfAssessments? Aren't they really representing the same thing? 🙂

File: DigitalLearningSolutions.Data.Migrations/Properties/Resources.resx [hunk 3]
```diff
@@ -121,6 +121,12 @@
   <data name="ApplyLPDefaultsSPChanges" type="System.Resources.ResXFileRef, System.Windows.Forms">
     <value>..\Scripts\ApplyLPDefaultsSPChanges.sql;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;utf-16</value>
   </data>
+  <data name="DLSV2_95_AddSystemVersioning" type="System.Resources.ResXFileRef, System.Windows.Forms">
+    <value>..\scripts\dlsv2-95_addsystemversioning.sql;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;Windows-1252</value>
+  </data>
+  <data name="DLSV2_95_RemoveSystemVersioning" type="System.Resources.ResXFileRef, System.Windows.Forms">
+    <value>..\scripts\dlsv2-95_removesystemversioning.sql;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;Windows-1252</value>
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Migrations/Properties/Resources.resx:128]: These need to have the same capitalisation as the files otherwise the build will fail 🙂

File: DigitalLearningSolutions.Web/Views/Shared/Error/Forbidden.cshtml [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
-  </div>
-</div>
+﻿@using DigitalLearningSolutions.Web.ViewModels.LearningPortal
+@model ErrorViewModel
+@{
+  ViewData["Title"] = "Learning Portal - You don't have permission to view this page";
+}
+
+<div class="nhsuk-grid-row">
+  <div class="nhsuk-grid-column-two-thirds nhsuk-u-reading-width">
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Web/Views/Shared/Error/Forbidden.cshtml:8]: The title above probably shouldn't reference the learning portal? 🙂

File: DigitalLearningSolutions.Web/Views/Shared/Error/PageNotFound.cshtml [hunk 5]
```diff
@@ -1,14 +1,14 @@
...
-    </div>
-</div>
+﻿@using DigitalLearningSolutions.Web.ViewModels.LearningPortal
+@model ErrorViewModel
+@{
+  ViewData["Title"] = "Learning Portal - Page not found";
+}
+
+<div class="nhsuk-grid-row">
+  <div class="nhsuk-grid-column-two-thirds nhsuk-u-reading-width">
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Web/Views/Shared/Error/PageNotFound.cshtml:8]: Same comment about the view data title 🙂

File: DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml [hunk 6]
```diff
@@ -23,7 +27,7 @@
   <link rel="icon" href="@Url.Content("~/favicon.ico")">
 
   <!-- Open Graph -->
-  <meta property="og:url" content="@Context.Request.GetDisplayUrl()">
+  @*<meta property="og:url" content="@Context.Request.GetDisplayUrl()">*@
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml:30]: Do we not need this? If so, probably worth removing instead of commenting out 🙂

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            Alter.Table("SelfAssessmentResults")
+                .AddColumn("SupportingComments").AsString(int.MaxValue).Nullable()
+                .AddColumn("VerificationRequested").AsDateTime().Nullable()
+                .AddColumn("VerificationOutcomeId").AsInt32().Nullable()
+                .AddColumn("VerifierAdminID").AsInt32().Nullable().ForeignKey("AdminUsers", "AdminID")
+                .AddColumn("VerifierComments").AsString(int.MaxValue).Nullable()
+                .AddColumn("VerifiedDate").AsDateTime().Nullable();
+            Execute.Sql(Properties.Resources.DLSV2_95_AddSystemVersioning);
+        }
+        public override void Down()
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:139]: Have you checked that this `Down` migration works? Worth double checking, just in case we need to undo the migration at any point. I think some of the existing `Down` migrations might not work at the moment, at least I was having some issues with them 🙂

File: DigitalLearningSolutions.Web/Startup.cs [hunk 8]
```diff
@@ -93,8 +93,8 @@ public void Configure(IApplicationBuilder app, IMigrationRunner migrationRunner,
                 app.UseBrowserLink();
             }
 
-            app.UseExceptionHandler("/LearningPortal/Error");
-            app.UseStatusCodePagesWithReExecute("/LearningPortal/StatusCode/{0}");
+            app.UseExceptionHandler("/Frameworks/Error");
+            app.UseStatusCodePagesWithReExecute("/Frameworks/StatusCode/{0}");
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Web/Startup.cs:97]: It's probably worth moving these two actions to their own controller, e.g. a LearningSolutionsController? It's probably also worth moving the AccessibilityHelp and Terms actions so that those footer links can be shared as well. What do you think? I'm happy to make that change if you prefer 🙂

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs [hunk 9]
```diff
@@ -0,0 +1,172 @@
+﻿namespace DigitalLearningSolutions.Data.Migrations
+{
+    using FluentMigrator;
+    using FluentMigrator.SqlServer;
+    [Migration(202011101344)]
+    public class CompetencyFrameworkDBChanges : Migration
+    {
+        public override void Up()
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:8]: I think these changes have slightly broken the service tests 🙂

**Review by FridaTveit — COMMENTED** [review]
Nice, thanks for sorting this! I've left a few comments and questions, let me know if any of it isn't clear or doesn't make sense! 🙂

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs (hunk 2 above)
Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:21]: Again - these were on another branch and somehow got dragged into the pull request, sorry :-(
Frameworks aren't the same as Self Assessments now that I understand it better. Self Assessments let a manager pull together competencies from one or more frameworks to present to a user. The frameworks are created and managed by a new framework developer role and I'm being asked to start work on the interface for that based on the LDA project outputs.

File: DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml (hunk 6 above)
Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml:30]: Unsure on this one, commented out because it was erroring. Is it important? - wasn't sure what it did tbh

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs (hunk 7 above)
Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:139]: will do 👍

File: DigitalLearningSolutions.Web/Startup.cs (hunk 8 above)
Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Web/Startup.cs:97]: Good point - I hadn't tested the AccessibilityHelp and Terms links but they don't work from elsewhere. It might be safer for you to make that change if you're happy to?

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs (hunk 9 above)
Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:8]: Yes - wasn't expecting them to be part of the pull request so hadn't checked that they were ready - sorry :(

File: DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml (hunk 6 above)
Comment (FridaTveit) [reply DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml:30]: It's for defining how the link appears when sharing on social media (https://css-tricks.com/essential-meta-tags-social-media/) so fine to remove if you don't want these pages to be shared on social media 🙂

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs (hunk 2 above)
Comment (FridaTveit) [reply DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:21]: Ah okay 🙂

File: DigitalLearningSolutions.Web/Startup.cs (hunk 8 above)
Comment (FridaTveit) [reply DigitalLearningSolutions.Web/Startup.cs:97]: Sure, happy to. Once this has been merged I'll do that and open a PR 🙂

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs (hunk 9 above)
Comment (FridaTveit) [reply DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:8]: If you don't want the framework changes to be included in this PR you can do an [interactive rebase](https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history) and remove the first two commits from this branch (looks like they contain the frameworks work). Just make sure you have those commits on another branch first (e.g. by creating a copy of this branch) so you don't lose them 🙂

File: DigitalLearningSolutions.Data.Migrations/Properties/Resources.resx [hunk 10]
```diff
@@ -122,10 +122,10 @@
     <value>..\Scripts\ApplyLPDefaultsSPChanges.sql;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;utf-16</value>
   </data>
   <data name="DLSV2_95_AddSystemVersioning" type="System.Resources.ResXFileRef, System.Windows.Forms">
-    <value>..\scripts\dlsv2-95_addsystemversioning.sql;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;Windows-1252</value>
+    <value>..\Scripts\DLSV2_95_AddSystemVersioning.sql;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;Windows-1252</value>
   </data>
   <data name="DLSV2_95_RemoveSystemVersioning" type="System.Resources.ResXFileRef, System.Windows.Forms">
-    <value>..\scripts\dlsv2-95_removesystemversioning.sql;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;Windows-1252</value>
+    <value>..\Scripts\DLSV2_95_RemoveSystemVersioning.sql;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;Windows-1252</value>
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Migrations/Properties/Resources.resx:128]: The build still isn't very happy with this. Should it be a hyphen in front of the 95 maybe, not an underscore? 🙂

Comment (FridaTveit) [reply DigitalLearningSolutions.Data.Migrations/Properties/Resources.resx]: The build is happy now 🎉 😄

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs (hunk 9 above)
Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:8]: I'm not confident doing this without losing the commits. I will make sure they are working with the service tests and leave them in here if that's okay?

Comment (FridaTveit) [reply DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:8]: Sure, that's fine 🙂

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs (hunk 7 above)
Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:139]: The Down() migration works, now (after minor headaches!) :)

File: DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs (hunk 9 above)
Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Data.Migrations/202011101344_CompetencyFrameworkDBChanges.cs:8]: Service tests all pass now :)

**Comment by kevwhitt-hee** [comment]
Happy for this to be merged, now?

**Review by FridaTveit — APPROVED** [review]
Looks great and looks like all the tests are working now 😄 Once this has been merged I'll move the error actions + terms & accessibility to their own controller 🙂
